### DISCUSSION TITLE
Constraint that validates a property of DTO is a collection of objects of a specific FQCN

### DISCUSSION THREAD

**Discussion by lbenz** [symfony/symfony#54358] [Q&A] [answered]
I use DTO since symfony 6.3 with `#[MapRequestPayload]` attribute and I could not find a way to perform a validation on a property that holds a collection of a specific class type. In the example below, I perform a POST API call with a payload that represents this DTO, and everytime I get an error.

Dto definition:
```
class MyDto
{
    public function __construct()
    {
        #[Assert\NotBlank]
        public int $id,
        
        #[Assert\All(
            new Assert\Type(EmbeddedObject::class),
        )]
        #[Assert\Valid]
        public array $listOfEmbeddedObject = [],
    }
}
```

Embedded property definition:
```
class EmbeddedObject
{
    public function __construct()
    {
        #[Assert\NotBlank]
        public string $name,
    }
}
```

Error:
```
{
    "type": "https://symfony.com/errors/validation",
    "title": "Validation Failed",
    "status": 422,
    "detail": "embeddedObject[0]: This value should be of type EmbeddedObject.",
    "violations": [
        {
            "propertyPath": "embeddedObject[0]",
            "title": "This value should be of type EmbeddedObject.",
            "template": "This value should be of type {{ type }}.",
            ...
```

The only way I found is to use an `Assert\Collection` validation on the DTO, but this way is not very reusable:
```
        #[Assert\All(
            new Assert\Collection([
                'name' => [
                    new Assert\Type('string'),
                ],
            ])
        )]
        public array $listOfEmbeddedObject = [],
```

**Comment by MatTheCat** [comment, marked as answer]
Your issue is that there is no way to know which type `MyDto::$listOfEmbeddedObject` items should be deserialized to. That’s why the `Type` constraint fails but the `Collection` one works: `$listOfEmbeddedObject` actually ends up as an array of arrays.

If your app depends on `phpdocumentor/reflection-docblock`,

```php
/** @var EmbeddedObject[] */
#[Assert\All(
    new Assert\Type(EmbeddedObject::class),
)]
#[Assert\Valid]
public array $listOfEmbeddedObject = [],
```

should work (not even sure the `Type` assertion is needed then).

**Reply by lbenz** [reply]
Wow thanks it works like a charm and the headashes go away 😃. By the way I confirm there is no need of `Assert\All`. Is there any documentation where we can find this ?

**Reply by MatTheCat** [reply]
Yay :grin: 

> there is no need of `Assert\All`

Note that it can still be useful in some cases; if you don’t use the serializer to instantiate the DTO e.g.

> Is there any documentation where we can find this ?

The closest match would be https://symfony.com/doc/current/components/serializer.html#recursive-denormalization-and-type-safety. Clicking the [PropertyInfo Component](https://symfony.com/doc/current/components/property_info.html) link, you can find the [PhpDocExtractor](https://symfony.com/doc/current/components/property_info.html#phpdocextractor) which seems the best fit in your case.

I agree it’s not very intuitive though..!
